Client/Frontend/Widgets/BottomSheetViewController.swift

```diff
@@ -0,0 +1,257 @@
+        backgroundOverlayView.snp.makeConstraints {
+            let bottomInset = parent?.view.safeAreaInsets.bottom ?? 0
+            $0.leading.trailing.top.equalToSuperview()
+            $0.bottom.equalToSuperview().offset(bottomInset)
+        }
+        
+        handleView.snp.remakeConstraints {
+            $0.width.equalTo(UX.handleWidth)
+            $0.height.equalTo(UX.handleHeight)
+
+            $0.centerX.equalTo(contentView)
+            $0.top.equalTo(contentView).offset((UX.handleMargin - UX.handleHeight) / 2)
+        }
+        
+        closeButton.snp.makeConstraints {
+            $0.top.equalToSuperview().inset(10)
+            $0.right.equalToSuperview().inset(7)
+            $0.size.equalTo(26)
+        }
+    }
+    
+    // MARK: - Animations
+    
+    @objc private func handlePanGesture(_ pan: UIPanGestureRecognizer) {
+        // The view shouldn't be draggable on iPads
+        if showAsPopup { return }
+
+        let translation = pan.translation(in: contentView)
+        yPosition += translation.y
+        pan.setTranslation(CGPoint.zero, in: contentView)
+
+        if pan.state != .ended { return }
+        
+        let projectedVelocity = project(initialVelocity: pan.velocity(in: contentView).y,
+                                        decelerationRate: UIScrollView.DecelerationRate.normal.rawValue)
+        
+        let nextYPosition: CGFloat
+        
+        let bottomHalfOfChildView = view.frame.height - (contentView.frame.height / 2)
+        let pannedPastHalfOfViewHeight = yPosition > bottomHalfOfChildView
+        
+        let closeByVelocity = projectedVelocity + yPosition > contentView.frame.maxY
+        
+        if pannedPastHalfOfViewHeight || closeByVelocity {
+            nextYPosition = view.frame.maxY
+        } else {
+            nextYPosition = 0
+        }
+
+        UIView.animate(withDuration: animationDuration, animations: {
+            self.yPosition = nextYPosition
+            
+        }) { _ in
+            if nextYPosition > 0 {
+                self.view.removeFromSuperview()
+                self.removeFromParent()
+            }
+        }
+    }
+
+    private func show() {
+        contentView.isHidden = false
+        UIView.animate(withDuration: animationDuration) {
+            self.yPosition = self.initialDrawerYPosition
+        }
+    }
+    
+    @objc private func closeView() {
+        close()
+    }
+
+    func close() {
+        UIView.animate(withDuration: animationDuration, animations: {
+            self.yPosition = self.view.frame.maxY
+        }) { _ in
+            self.view.removeFromSuperview()
+            self.removeFromParent()
+        }
+    }
+    
+    // Distance travelled after decelerating to zero velocity at a constant rate (credit: a WWDC video)
+    private func project(initialVelocity: CGFloat, decelerationRate: CGFloat) -> CGFloat {
+      return (initialVelocity / 1000.0) * decelerationRate / (1.0 - decelerationRate)
```

Comment:
Would this break if `decelerationRate == 1.0`? (e.g. ` ... / (1.0 - 1.0)`?

Reply:
cc @kylehickinson

Reply:
Yeah but a deceleration rate of 1.0 wouldn't make sense (would mean no deceleration). We can add a check if you want though